### PR TITLE
Detect_sharp_edges_plugin works with a Scene_surface_mesh_item

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes
The Detect_sharp_edges_plugin now works with SM_items. 

To do so, the file Detect_features_in_polyhedra.h has been BGL-ized and a new property_map has been added to Polyhedron_type, Scene_surface_mesh_item and Mesh_polyhedron_3. 

This property map contains the feature halfedges.
## Release Management

This PR is based on #1983 

* Affected package(s):Polyhedron_demo, Polygon_mesh_processing
* Issue(s) solved (if any): fix #1931

### CODE REVIEW

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+}
+  
+// -----------------------------------
+// Private methods
+// -----------------------------------
+template <typename P_, typename I_>
+typename Detect_features_in_polyhedra<P_, I_>::Vector_3
+Detect_features_in_polyhedra<P_, I_>::
+facet_normal(const Polyhedron& polyhedron, const face_descriptor& f) const
+{
```
Comment (janetournois) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h]: you can replace this function by a simple call to `PMP::compute_facet_normal` everywhere in the file

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    ++current_surface_index_;
+  }
+}
+
+
+template <typename P_, typename I_>
+void
+Detect_features_in_polyhedra<P_, I_>::
+detect_vertices_incident_patches(Polyhedron& polyhedron)
+{
```
Comment (janetournois) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h]: is it possible to have a `const Polyhedron&` here?

File: BGL/include/CGAL/boost/graph/helpers.h [hunk 3]
```diff
@@ -751,6 +751,20 @@ void clear(FaceGraph& g)
   CGAL_postcondition(num_faces(g) == 0);
 }
 
+
+template <typename FaceGraph>
+typename boost::graph_traits<FaceGraph>::degree_size_type
+face_degree(typename boost::graph_traits<FaceGraph>::halfedge_descriptor hd, const FaceGraph& g)
```
Comment (janetournois) [inline BGL/include/CGAL/boost/graph/helpers.h]: this function should be documented in the BGL helper functions

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+#include <CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra_fwd.h>
+#include <CGAL/Compare_handles_with_or_without_timestamps.h>
+#include <CGAL/Polygon_mesh_processing/compute_normal.h>
+#include <set>
+
+namespace CGAL {
+namespace Polygon_mesh_processing {
+
+  template <typename Polyhedron, typename FT, typename PatchId_pmap>
+void detect_features(Polyhedron& p,
```
Comment (janetournois) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h]: This function should be documented, and use named parameters. I will add some hints in the main discussion

**Comment by janetournois** [comment]
To document `PMP::detect_features`, I would suggest to add named parameters for
- `SharpEdgesPmap`, inspired by [`PMP_edge_is_constrained_map`](http://doc.cgal.org/latest/Polygon_mesh_processing/group__namedparameters.html#PMP_edge_is_constrained_map)
- [`Face_patch_map`](http://doc.cgal.org/latest/Polygon_mesh_processing/group__namedparameters.html#PMP_face_patch_map) replacing `PatchId_pmap`
- a property map for `vertex_incident_patches`

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h (hunk 2 above)
Comment (maxGimeno) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h]: I don't think so, as we put in the PMAP in the end

Comment (maxGimeno) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/Detect_features_in_polyhedra.h]: I don't think so as we need to put in the PMAP in the end.

**Comment by lrineau** [comment]
It seems that this PR will require a small feature.

**Comment by afabri** [comment]
The class `Detect_features_in_polyhedra` must not be constructed with a patch id property map, as for each of its member functions the property map for the polyhedron that is parameter of the function.  The member functions should have the property map as additional parameter with a default.   
@maxGimeno  please have a look at #f3f0a37145fd6ff515e9990c0b1f4a189450e353 and finalize what I've started .

**Comment by maxGimeno** [comment]
It is already in master. I guess It has been done in #1983.

**Pull request closed without merging**